testssl.sh

```diff
@@ -7354,6 +7361,249 @@ tls_time() {
+     fi
+     out "$spaces"
+
+     if "$policy_ok"; then
+          if [[ $policy_mode == testing ]]; then
+               out "\"none\" is a valid policy but why are you using it?"
+               fileout "${jsonID}_policy" "INFO" "none is valid but not a helpful policy  at https://mta-sts.$domain/.well-known/mta-sts.txt"
+          elif [[ $policy_mode == testing ]]; then
+               out "valid but not enforced"
+               fileout "${jsonID}_policy" "INFO" "valid but not enforced policy  at https://mta-sts.$domain/.well-known/mta-sts.txt"
+          else
+               pr_svrty_good "valid and enforced"
+               fileout "${jsonID}_policy" "OK" "valid and enforced policy  at https://mta-sts.$domain/.well-known/mta-sts.txt"
+          fi
+          outln " policy https://mta-sts.$domain/.well-known/mta-sts.txt"
+     elif [[ -z "$policy" ]]; then
+          pr_svrty_low "no policy"
+          outln " at https://mta-sts.$domain/.well-known/mta-sts.txt"
+          fileout "${jsonID}_policy" "LOW" "no policy at https://mta-sts.$domain/.well-known/mta-sts.txt"
+     else
+          # missing: too short, not enforced, etc..
+          pr_svrty_low "invalid policy"
+          failreason_policy=( "${failreason_policy[@]/%/ | }" )
+          fileout "${jsonID}_policy" "LOW" "invalid policy  https://mta-sts.$domain/.well-known/mta-sts.txt: $(printf '%s' ${failreason_policy[@]})"
+          outln " at https://mta-sts.$domain/.well-known/mta-sts.txt: $(printf '%s' ${failreason_policy[@]})"
+     fi
+     out "$spaces"
+
+     if "$smtp_tls_record_ok"; then
+          outln "found (optional) TLS RPT TXT record '$smtp_tls_record'"
+          fileout "${jsonID}_tlsrpt" "INFO" "optional TLS-RPT TXT record '$smtp_tls_record'"
+     else
+          outln "No TLS RPT record"
+          fileout "${jsonID}_tlsrpt" "INFO" "no or invalid (optional) TLS RPT record '$smtp_tls_record'"
+     fi
+
+     return 0
+}
+
+# e.g. for removing the HTTP header. To be moved to the top
+#
+print_after_blankline() {
+     local first=true
+     local line=""
+
+     while read -r line; do
+          if ! "$first"; then
+               safe_echo "$line\n"
+          else
+               # ignore everything until we hit an empty line or a line with a blank or a CR / LF
+               if [[ -z "$line" ]] || [[ "$line" =~ ^[[:space:]]$ ]]; then
+                    first=false
+                    continue
+               fi
+          fi
+     done <<< $1
+}
+
+
+
+# RFC 6394
+# RFC 6698
+# RFC 7218
+# RFC 7671
+# RFC 7672
+# RFC 7673
+sub_dane() {
+     local tlsa_record=""
+     local rrsig_record=""
+     local spaces="$1"
+
+     # Not yet implemeted
+     return 0
+
+     pr_bold " DANE / DNSSEC                "
+
+     tlsa_record="$(get_tlsa_record _$PORT._tcp.$NODE)"
+     # parsing TLSA certificate usage, TLSA selector, TLSA matching type, hash
+     rrsig_record="$(get_rrsig_record $NODE)"
```

Comment:
To determine whether DANE is securely implemented:

- Determine whether the destination domain's MX RRset is DNSSEC signed and validates.  Best done by inspecting the "AD" bit from the local (loopback) DNS resolver.  If the "AD" bit is on for ". IN SOA ?" and off for "root-servers.net. IN SOA ?", then your local resolver appears to support DNSSEC, and "example.com. IN MX ?" should indicate whether the domain is signed.
- For each MX host determine whether either its A or AAAA RRset is DNSSEC signed (if either is non-existent, NODATA, the denial of existence may be insecure due to NSEC3 opt-out).
- If the A or AAAA RRset of, say, "smtp.example." is obtained via CNAME expansion, and AD=0, the zone may still be signed, check for the AD bit of "smtp.example. IN CNAME ?".
- If any of the MX hosts are in unsigned zones, at best "partial DANE support" is possible, which is subject to active downgrade to the unsigned MX hosts, by blocking connections to the signed ones.
- For each MX host in a signed zone, check for TLSA RRs:
    -  first with the base domain set to the CNAME expansion, if AD=1 for the address records despite CNAME expansion.
    - next with the base domain set to the original MX host name if no CNAME was used, was insecure, or no TLSA RRs found there.
- To check for TLSA RRs prepend "_25._tcp." to the base domain and perform a lookup with type "TLSA".
- OpenSSL `s_client` can be used to perform a DANE-validated STARTTLS session with given TLSA RRDATA, see the manpage.
- Alternatively, the OpenSSL API documentation for DANE can be found at: <https://www.openssl.org/docs/man1.1.1/man3/SSL_CTX_dane_enable.html>
- The "posttls-finger" program included with the Postfix source distribution, and in many OS Postfix distribution can perform DANE-based verification, though the output would need to be parsed to determine verification success, there's no exit code that directly reflects DANE authentication status.
- See also https://github.com/baknu/DANE-for-SMTP/wiki/2.-Implementation-resources